### PR TITLE
[Core] Fixing inconsistency in NR strategy

### PR DESCRIPTION
**Pull request by loumalouomega**
The NR holds a linear solver pointer, but the B&S already holds one. This PR removes that member variable and adds a check in the constructor

### CODE REVIEW

**Comment by loumalouomega** [comment]
Anyone?

**Comment by philbucher** [comment]
all those changes are delicate, see [my comment](https://github.com/KratosMultiphysics/Kratos/pull/6860#issuecomment-622937514)
I don't feel comfortable just approving this

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 1]
```diff
@@ -195,12 +193,21 @@ class ResidualBasedNewtonRaphsonStrategy
...
 
         // Tells to the Builder And Solver if the system matrix and vectors need to
         //be reshaped at each step or not
-        GetBuilderAndSolver()->SetReshapeMatrixFlag(mReformDofSetAtEachStep);
+        p_builder_and_solver->SetReshapeMatrixFlag(mReformDofSetAtEachStep);
+
+        // We check if the linear solver considered for the builder and solver is consistent
+        auto p_linear_solver = p_builder_and_solver->GetLinearSystemSolver();
+        if (p_linear_solver != pNewLinearSolver) {
+            KRATOS_WARNING("ResidualBasedNewtonRaphsonStrategy") << "Inconsistent linear solver in strategy and builder and solver. Considering the linear solver assigned to builder and solver :\n" << p_linear_solver->Info() << "\n instead of:\n" << pNewLinearSolver->Info() << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]: this should be an error I would say

**Review by philbucher — COMMENTED** [review]
to me this PR makes sense, thx for cleaning

can you please make a new constructor that does not accept the linear solver and deprecate the old one (the old one internally calls the new one to reduce code duplication if possible)

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]: Opinion @RiccardoRossi ?

**Comment by loumalouomega** [comment]
Done @philbucher

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 2]
```diff
@@ -338,6 +374,37 @@ class ResidualBasedNewtonRaphsonStrategy
...
+        : ResidualBasedNewtonRaphsonStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(rModelPart, pScheme, pNewConvergenceCriteria, pNewBuilderAndSolver, Settings)
+    {
+        KRATOS_TRY
+
+        // Getting builder and solver
+        auto p_builder_and_solver = GetBuilderAndSolver();
+
+        // We check if the linear solver considered for the builder and solver is consistent
+        auto p_linear_solver = p_builder_and_solver->GetLinearSystemSolver();
+        KRATOS_WARNING_IF("ResidualBasedNewtonRaphsonStrategy", p_linear_solver != pNewLinearSolver) << "Inconsistent linear solver in strategy and builder and solver. Considering the linear solver assigned to builder and solver :\n" << p_linear_solver->Info() << "\n instead of:\n" << pNewLinearSolver->Info() << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:403]: I think this should be an error, it is a significant problem in the input

**Comment by loumalouomega** [comment]
If I change it to an error I need to update many inconsistent tests (I
already tried and many tests fail). With a warning I can keep this PR
clean. We can change it to an error in a future PR and update the
problematic tests then

**Comment by philbucher** [comment]
> If I change it to an error I need to update many inconsistent tests (I
already tried and many tests fail). With a warning I can keep this PR
clean. We can change it to an error in a future PR and update the
problematic tests then

I expected nothing less :D

**Review by philbucher — APPROVED** [review]
thanks :+1:

**Pull request merged**